src/bernsteinyang.rs

```diff
@@ -0,0 +1,351 @@
+/// 
+/// The public methods of this type receive and return unsigned big integers as arrays of
+/// 64-bit chunks, the ordering of which is little-endian. Both the modulus and the integer
+/// to be inverted should not exceed 2 ^ (B * (L - 1) - 2)
+pub struct BYInverter<const B:usize, const L:usize> {
+    /// Modulus
+    modulus: ChunkInt<B,L>,
+
+    /// Adjusting parameter
+    adjuster: ChunkInt<B,L>,
+
+    /// Multiplicative inverse of the modulus modulo 2^B
+    inverse: i64
+}
+
+/// Type of the Bernstein-Yang transition matrix multiplied by 2^B
+type Matrix = [[i64; 2]; 2];
+
+impl<const B:usize, const L:usize> BYInverter<B,L> {  
+    fn step(f: &ChunkInt<B,L>, g: &ChunkInt<B,L>, mut delta: i64) -> (i64, Matrix) {
+        let (mut steps, mut f, mut g) = (B as i64, f.lowest() as i64, g.lowest() as i128);
+        let mut matrix: Matrix = [[1, 0], [0, 1]];
+
+        loop {
+            let zeros = steps.min(g.trailing_zeros() as i64);
+            (steps, delta, g) = (steps - zeros, delta + zeros, g >> zeros);
+            matrix[0] = [matrix[0][0] << zeros, matrix[0][1] << zeros];
+    
+            if steps == 0  { break; }      
+            
+            if delta > 0 {
+                (delta, f, g) =  (-delta, g as i64, -f as i128);
+                (matrix[0], matrix[1]) =  (matrix[1], [-matrix[0][0], -matrix[0][1]]);
+            }
+
+            let mask = (1 << steps.min(1 - delta).min(4)) - 1;
+            let w = (g as i64).wrapping_mul(f.wrapping_mul(3) ^ 12) & mask;
+            
+            matrix[1] = [matrix[0][0] * w + matrix[1][0], matrix[0][1] * w + matrix[1][1]];
+            g += w as i128 * f as i128;
+        }
+    
+        (delta, matrix)
+    }
+    
+    fn fg(f: ChunkInt<B,L>, g: ChunkInt<B,L>, matrix: Matrix) -> (ChunkInt<B,L>, ChunkInt<B,L>) {
+        ((matrix[0][0] * &f + matrix[0][1] * &g).shift(), (matrix[1][0] * &f + matrix[1][1] * &g).shift()) 
+    }
+
+    fn de(&self, d: ChunkInt<B,L>, e: ChunkInt<B,L>, matrix: Matrix) -> (ChunkInt<B,L>, ChunkInt<B,L>) {
+        let mask = ChunkInt::<B,L>::MASK as i64;
+        let mut md = matrix[0][0] * d.is_negative() as i64 + matrix[0][1] * e.is_negative() as i64;
+        let mut me = matrix[1][0] * d.is_negative() as i64 + matrix[1][1] * e.is_negative() as i64;
+
+        let cd = matrix[0][0].wrapping_mul(d.lowest() as i64).wrapping_add(matrix[0][1].wrapping_mul(e.lowest() as i64)) & mask; 
+        let ce = matrix[1][0].wrapping_mul(d.lowest() as i64).wrapping_add(matrix[1][1].wrapping_mul(e.lowest() as i64)) & mask;
+
+        md -= (self.inverse.wrapping_mul(cd).wrapping_add(md)) & mask;
+        me -= (self.inverse.wrapping_mul(ce).wrapping_add(me)) & mask;
+        
+        let cd = matrix[0][0] * &d + matrix[0][1] * &e + md * &self.modulus;
+        let ce = matrix[1][0] * &d + matrix[1][1] * &e + me * &self.modulus;
+
+        (cd.shift(), ce.shift())   
+    }
+
+    fn norm(&self, mut value: ChunkInt<B,L>, negate: bool) -> ChunkInt<B,L> {
```

Comment:
This should indicate the input range and output range

AFAIK: 

> Compute a = sign*a (mod M)
>
> with a in range (-2*M, M)
> result in range [0, M)

also why is the value mutated and returned as well?

Reply:
Agree. I was going to comment this part of the code.

> also why is the value mutated and returned as well?

It is mutable in order to mutate it within the function without introducing a new temporary variable.
It is returned, because the function takes the ownership of the argument instead of borrowing it mutably.
It is returned instead of being changed using a mutable reference because I want all function to be pure: https://en.wikipedia.org/wiki/Pure_function